File: packages/lynx-dev-client/ios/Sources/SigxDevClient/DevTemplateResourceFetcher.swift
import Foundation
import Lynx

/// Handles template and HMR hot-update fetching for LynxView.
/// Required for HMR — the template provider only handles initial loads.
class DevTemplateResourceFetcher: NSObject, LynxTemplateResourceFetcher {

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        return URLSession(configuration: config)
    }()

    func fetchTemplate(
        _ request: LynxResourceRequest,
        onComplete callback: @escaping (LynxTemplateResource?, Error?) -> Void
    ) {
        let url = request.url ?? ""

        if url.hasPrefix("file://") || (!url.hasPrefix("http://") && !url.hasPrefix("https://")) {
            fetchFromAssets(url.replacingOccurrences(of: "file://", with: ""), callback: callback)
            return
        }

        guard let requestURL = URL(string: url) else {
            callback(nil, NSError(domain: "com.sigx.devclient", code: 400,
                                  userInfo: [NSLocalizedDescriptionKey: "Invalid URL: \(url)"]))
            return
        }

        session.dataTask(with: requestURL) { data, response, error in
            if let error = error {
                callback(nil, error)
                return
            }
            guard let data = data else {
                callback(nil, NSError(domain: "com.sigx.devclient", code: 404,
                                      userInfo: [NSLocalizedDescriptionKey: "No data received"]))
                return
            }
            let resource = LynxTemplateResource(nsData: data)
            callback(resource, nil)
        }.resume()
    }

    func fetchSSRData(
        _ request: LynxResourceRequest,
        onComplete callback: @escaping (Data?, Error?) -> Void
    ) {
        // SSR not used in dev mode; return nil
        callback(nil, nil)
    }

    private func fetchFromAssets(
        _ path: String,
        callback: @escaping (LynxTemplateResource?, Error?) -> Void
    ) {
        guard let bundlePath = Bundle.main.path(forResource: path, ofType: nil) else {
            callback(nil, NSError(domain: "com.sigx.devclient", code: 404,
                                  userInfo: [NSLocalizedDescriptionKey: "Asset not found: \(path)"]))
            return
        }

        do {
            let data = try Data(contentsOf: URL(fileURLWithPath: bundlePath))
            let resource = LynxTemplateResource(nsData: data)
            callback(resource, nil)
        } catch {
            callback(nil, error)
        }
    }
}

File: packages/lynx-dev-client/android/src/main/kotlin/com/sigx/devclient/DevTemplateResourceFetcher.kt
package com.sigx.devclient

import android.content.Context
import android.util.Log
import com.lynx.tasm.resourceprovider.LynxResourceCallback
import com.lynx.tasm.resourceprovider.LynxResourceRequest
import com.lynx.tasm.resourceprovider.LynxResourceResponse
import com.lynx.tasm.resourceprovider.template.LynxTemplateResourceFetcher
import com.lynx.tasm.resourceprovider.template.TemplateProviderResult
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Handles template and HMR hot-update fetching for LynxView.
 * Required for HMR to work — the AbsTemplateProvider only handles initial loads.
 */
@Suppress("UNCHECKED_CAST")
class DevTemplateResourceFetcher(context: Context) : LynxTemplateResourceFetcher() {

    private val appContext: Context = context.applicationContext

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    override fun fetchTemplate(
        request: LynxResourceRequest,
        callback: LynxResourceCallback<TemplateProviderResult>
    ) {
        if (request == null) {
            callback.onResponse(failed("request is null"))
            return
        }

        val url = request.url
        if (url.startsWith("file://") || (!url.startsWith("http://") && !url.startsWith("https://"))) {
            fetchFromAssets(url.removePrefix("file://"), callback)
            return
        }

        val httpRequest = Request.Builder().url(url).build()
        httpClient.newCall(httpRequest).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (it.isSuccessful && it.body != null) {
                        val bytes = it.body!!.bytes()
                        val result = TemplateProviderResult.fromBinary(bytes)
                        callback.onResponse(LynxResourceResponse.onSuccess(result))
                    } else {
                        callback.onResponse(failed("HTTP ${it.code}: ${it.message}"))
                    }
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                Log.e("SigxDevClient", "Template resource fetch failed: ${request.url}", e)
                callback.onResponse(failed(e))
            }
        })
    }

    override fun fetchSSRData(
        request: LynxResourceRequest,
        callback: LynxResourceCallback<ByteArray>
    ) {
        if (request == null) {
            callback.onResponse(failed("request is null"))
            return
        }

        val httpRequest = Request.Builder().url(request.url).build()
        httpClient.newCall(httpRequest).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (it.isSuccessful && it.body != null) {
                        callback.onResponse(LynxResourceResponse.onSuccess(it.body!!.bytes()))
                    } else {
                        callback.onResponse(failed("HTTP ${it.code}: ${it.message}"))
                    }
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                callback.onResponse(failed(e))
            }
        })
    }

    private fun fetchFromAssets(
        path: String,
        callback: LynxResourceCallback<TemplateProviderResult>
    ) {
        try {
            val bytes = appContext.assets.open(path).use { it.readBytes() }
            val result = TemplateProviderResult.fromBinary(bytes)
            callback.onResponse(LynxResourceResponse.onSuccess(result))
        } catch (e: Exception) {
            callback.onResponse(failed(e))
        }
    }

    private fun <T> failed(message: String): LynxResourceResponse<T> =
        LynxResourceResponse.onFailed(Throwable(message)) as LynxResourceResponse<T>

    private fun <T> failed(e: Throwable): LynxResourceResponse<T> =
        LynxResourceResponse.onFailed(e) as LynxResourceResponse<T>
}
